src/main/frontend/sections/Toolbar.tsx

```diff
@@ -16,8 +16,52 @@ export const Toolbar: FC = () => {
 	const scriptContext = useContext(ScriptContext);
 	const script = scriptContext.draft;
 
+	const samples: Sample[] = fetchSamples();
 	const [, copy] = useCopyToClipboard();
 
+	function fetchSamples() {
+		try {
+			const environmentContext = useContext(EnvironmentContext);
+			const dataSampleScripts = environmentContext.sampleScripts;
+			if (Array.isArray(dataSampleScripts)) {
+				const additionalSamples = dataSampleScripts.map(item => ({
+					label: item['label'] ?? 'Sample',
+					config: JSON.parse(item['configJson'] ?? '{}') as Script,
+				}));
+				return [...SAMPLES, ...additionalSamples];
```

Comment:
You’re doing the parsing and construction of the array on every render. Since the list never changes, wrap the call to `fetchSamples` in [`useMemo`](https://react.dev/reference/react/useMemo):

```ts
const samples = useMemo(fetchSamples, [environmentContext.sampleScripts]);
```

(note that this requires you to move the `useContext` call outside the `fetchSamples`, as suggested in my other comment).